File: #154_Find_minimum_in_Rotated_Sorted_Array_II/Solution.java
public class Solution {
    public int findMin(int[] nums) {
    	int len = nums.length;
    	int lo = 0, hi = len - 1;
    	while (true) {
    		/* inorder */
    		if (nums[lo] < nums[hi]) return nums[lo];
    		if (lo == hi) return nums[lo];
    		if (lo == hi - 1) return Math.min(nums[lo], nums[hi]);
    		int mid = (lo + hi) / 2;
			if (nums[lo] < nums[mid]) lo = mid + 1; // search right
    		else if (nums[lo] > nums[mid]) hi = mid; // search left
    		else lo++; // worst case O(n)
    	}
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {10,1,10,10,10};
    	System.out.println(sol.findMin(t));
    }
}

File: #154_Find_minimum_in_Rotated_Sorted_Array_II/Solution.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def findMin(self, nums):
        if len(nums) == 1:
        	return nums[0]
        if len(nums) == 2:
        	return min(nums)
        if nums[0] < nums[len(nums) - 1]:
        	return nums[0]
        mid = (len(nums) - 1) / 2
        if nums[mid] > nums[0]: # left is sorted, search right
        	return self.findMin(nums[mid + 1:])
        if nums[mid] < nums[0]: # left is unsorted, search left
        	return self.findMin(nums[:mid + 1])
        if nums[mid] < nums[len(nums) - 1]: # right is sorted, search left
            return self.findMin(nums[:mid + 1])
        if nums[mid] > nums[len(nums) - 1]:# right is un sorted, search right
            return self.findMin(nums[mid + 1:])
        return self.findMin(nums[1:])

sol = Solution()
print sol.findMin([1, 1, 2, 2, 2, 0])